base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1420,19 +1420,26 @@ fn find_orphan_chain_tips<T: BlockchainBackend>(
 ) -> Vec<BlockHash>
 {
     let mut tip_hashes = Vec::<BlockHash>::new();
+    let mut next_parent_hash = None;
+    let mut next_parent_height = 0;
     db.for_each_orphan(|pair| {
         let (_, block) = pair.unwrap();
         if (block.header.prev_hash == parent_hash) && (block.header.height == parent_height + 1) {
-            let next_parent_hash = block.hash();
-            let mut orphan_chain_tips = find_orphan_chain_tips(db, block.header.height, next_parent_hash.clone());
-            if !orphan_chain_tips.is_empty() {
-                tip_hashes.append(&mut orphan_chain_tips);
-            } else {
-                tip_hashes.push(next_parent_hash);
-            }
+            // we found a match, let save to call later
+            next_parent_hash = Some(block.hash());
```

Comment:
Right now, this doesn't do the same thing as the code on the left. If there are 2 or more orphan blocks that are parents (their prev_hash == parent_hash), this loop only identifies the last one.

If you  return all of them in a vec, you can proceed to the next part, which reduces the read accessor problem by making this BFS rather than DFS, but still leaves a recursive call, which might still lead to problems down the line, but it is decidedly an improvement.